File: three.pas
(* three.pas 		    *)
(* Three Card Game Paradox  *)
(* by Conor Gilmer          *)

program three;

var 
winning, chosen, remove:integer;
stick, twist, games, num, numRand, numA, numB, numC:cardinal;
stickR, twistR, gamesR:real;
stickpc, twistpc:real;
letter:char;

(* return a random card a number between 1 and 3 *)
function randomcard(cards:integer) : integer;
var
rvar:integer;
begin
	numRand := numRand +1;
	randomize;
	rvar := random(cards) + 1;	
(*	writeln('random = ', rvar);*)

	if rvar = 1 then 
	begin
		numA := numA +1;
	end
	else if rvar = 2 then
	begin
		numB := numB +1;
	end
	else if rvar = 3 then
	begin
		numC := numC +1;
	end
	else
		writeln('randomizer malfunctioned');

	randomcard := rvar;
end;

(* play a game of selecting from three cards *)
procedure play;
begin

	winning := randomcard(3);
	chosen := randomcard(3);
	
	repeat
  	remove := randomcard(3);
	until (remove <> winning) and (remove <> chosen);

	writeln('Game: ', num,' - winning = ', winning, ' | chosen = ', chosen, ' | remove = ', remove );

	if winning = chosen then
	begin
		stick := stick + 1;
	end
	else
		twist := twist + 1;
end;

(* Main Program Three Card Paradox *)
begin

	games := 100;

	writeln('Three Card Game Paradox');
	for num :=1 to games do
		play;
	writeln('Results:');
	stickR := stick;
	twistR := twist;
	gamesR := games;
	stickpc := (stickR/gamesR) *100.0;
	twistpc := (twistR/gamesR) *100.0;
	writeln('Game played ', games, ' times');
	writeln('Sticking wins ', stick, ' times (', stickpc:2:2, '%)');
	writeln('Changing wins ', twist, ' times (', twistpc:2:2, '%)');
	writeln('Verdict:');
	if stick < twist then
	begin
		writeln('Twisting wins more!');
	end
	else
		writeln('Sticking wins more!');
	writeLn('The End.');

	writeln;
	writeln('How Random is the Randomizer?');
	writeln('Randomizer called ', numRand, ' times');
	writeln('A chosen ', numA, ' times(',((numA/numRand)*100):2:2,'%)');
	writeln('B chosen ', numB, ' times(',((numB/numRand)*100):2:2,'%)');
	writeln('C chosen ', numC, ' times(',((numC/numRand)*100):2:2,'%)');
	
	writeLn('End of Randomizer analysis.');
end.

File: three.f
c three.f
c Three Card Game Paradox
c by Conor Gilmer
c
      program three
c variables
      integer stick, twist, games, outcome, randcalled, a, b, c
      real stickpc, twistpc
      integer run
      common randcalled, a, b, c
     
      stick = 0
      twist = 0
      games = 250
      randcalled = 0
      a = 0
      b = 0
      c = 0

      print *,"Three Card Game Paradox!"
      print *,"Play ", games, " Times"
c run the game games times
      do i=1, games
        outcome = run(i)
        if (outcome == 1) then
                stick = stick + 1
                print *, "Sticking Wins"
        else
                twist = twist + 1
                print *, "Twisting Wins"
        end if
      end do
c output the results
      twistpc = (real(twist)/real(games)) * 100
      stickpc = (real(stick)/real(games)) * 100
      print *,"Results:"
      print *,"Sticking wins ", stick, " times (",stickpc, " %)."
      print *,"Changing wins ", twist, " times (",twistpc, " %)."
      if (stick.le.twist) then
        print *,("Changing is better")
      else
        print *,("Sticking is better")
      endif
      print *,"The End!"

      print *,"How Random is the irand() function?"
      print *,"Random function called ", randcalled
      print *,"A ", a, " times(", (real(a)/real(randcalled))*100, "%)"
      print *,"B ", b, " times(", (real(b)/real(randcalled))*100, "%)"
      print *,"C ", c, " times(", (real(c)/real(randcalled))*100, "%)"
      print *,"End of Random function analysis"
      end program three

c function to generate a simple random number between 1 and 3
c modulus means its numbers between 0 and 3 but if zero random again
      integer function simplerandom(x)
      integer x, randcalled, a, b, c
      common randcalled, a, b, c

c     do
        randcalled = randcalled + 1
        simplerandom = mod(irand(),x)
        if (simplerandom.eq.0) then
                a = a + 1;
        else if (simplerandom.eq.1) then
                b = b + 1;
        else if (simplerandom.eq.2) then
                c = c + 1;
        else 
                c = c;
        endif
c        print*,"Randcalled ", randcalled, "  returning ", simplerandom
c        if (simplerandom.ne.0) exit
c    end do
      return
      end function

c This is a function to play a game and return if stick or twisting won
      integer function run(x)
      integer x
      integer winning, chosen, remove
      character cards(3)
      integer simplerandom
      cards(1) = "A"
      cards(2) = "B"
      cards(3) = "C"
      winning = simplerandom(3)    
c      remove = simplerandom(3)    
      chosen = simplerandom(3)   

      do
       remove = simplerandom(3)    
       if ((remove.ne.winning) .and. (remove.ne.chosen)) exit
      end do
      print *,"winning ",cards(winning+1)," chose ",cards(chosen+1)
      print *,"remove ", cards(remove+1)

      if (winning.eq.chosen) then
        run = 1
      else
        run = 0
      endif
      RETURN 
      END
